README.md
To run the server, 

Pre-Requisites to make sense of the server:

- Have eXe installed

run this to sync your database


cd ustadmobil-code-umclouddjango/
$ python manage.py syncdb

cd ustadmobil-code-umclouddjango
$ python manage.py runserver

The server will run on 127.0.0.1 over port 8000 or you can specify the port and IP after runserer. 

To install date-time
sudo pip install django-datetime-widget



releasenotes.txt
v0.7.9b
- Added ability for superuser to log in without external approval checks.

v0.7.9
- Fixed unit tests
- minor tweaks for usability
- Fixed build bug01
- Fixed build bug02

v0.7.5
- Fixed log out bug

v0.7.4
- Accept approval of users: just change to make final screen as 'confirmation' with a reassuring checkbox or something of the like (can't be reasonably confused with an error message).
- Users: make sure to show extra fields in the edit user section.  Add textfield 'notes' (multi line extended length text field - only visible on edit user)
- New Class screen: change 'associative school' to just 'school'.  Remove link to 'create an associative school first'.  Remove link to 'create a new user'.
- Picklists - check icon size (e.g. http://umcloud1.ustadmobile.com:8010/allclassnew/ new class screen).  Check occurrences of class picklist use the icon which is three users instead of black boxes.
- Add organization code at the top for admin - so they know what code to give out.
- Add green confirmation bar on success of save student, school, etc.
- Messages functionality hidden. 
-Bug Fixes and Organisational and initial Role based filtering.

v0.7.3
-Developed and added messages to message users and made it look fairly decent but not mordern yet.

v0.7.2
- Registration and Sign up divided into two.

v0.7.1
-Added organisation management for staff/super users
-Polish the User Requests page.

v0.7.0
-Eliminate ID from table list
-Email is not mandatory
-Phone number mandatory
-Added field to user's profile: date of birth
-Added field to user's profile: gender, location
-Added "Add another" on all creation forms.
-Basic implementation of confirmations.
-First Name and Last name are mandatory
-Search in picklist is against all contains and not just starts with
-Packages called blocks.
-Toolbar on top has blocks and blocks view created
-User Approval: Intuitive way to accept and reject users 
-Icons are coloured
-Manage Blocks
-Save terminology and Gramatical errors fixed
-Added doublt type password fields for sigups


v0.6.2
-Updated picklists to filter and search for contains instead of starts-with

v0.6.1
-89% unit test code coverage updates.

v0.6.0
-82% unit test code coverage updates.

v0.5.9.1
-Fixed unit test bug to account for test files not present (exe elp file)

v0.5.9
-Added nother option in user_new to submit user creation and create another which redirects to another form.
-Got rid of ID fields in most tables but retained number in first field (usually name)
-Make pick list of user approval for org manager
-Make course pic black and white too to avoid confusion.
-Made First Name , Last Name, DOB fields required in register (website)
-Fixed bug in pending requests check for non organisation managers.
-Made First Name, Last Name, UserName, fields mandatory in New User
-Re ordered the Toolbar at the top to be Users, Schools, Classes, Courses
-Renamed Packages to Blocks (to be more in line with how other LMSs name course blocks)
-Disabled grunt testing on course upload.
-Updated user creation (internal) and website unit tests.
-Added more tests to unit testing including file (block/elp/package/exe) upload
-More updates to unit testing

v0.5.8
-Renamed Managemewnt's Manage X to X
-Removed duplicate new X / Add New buttons
-Removed extra text on pages.
-Fixed bug in creating class when no teachers defined.
-Changed icons to show classes are 3 users and users as 1 user (icon)
-Included a navigation panel on Management on top.
-Sign in loads to Users Management.
-Added Hover effect for selected menu.

v0.5.7
-Fixed course creation bug: It was still looking for price in code. Fixed it and model's updated to default to 0
-Seperated logic and UI for uploading, exe export and grunt tests

v0.5.6
-Added 404.html and 500.html pages. These get active when Debug is False
-Updated umclouddj-server-tar-deployment.sh to set Debug as False and add allowed hosts to the server in production.
-Added organisational constraints to Allclasses's school field.
-Modified Class to return name on model's object call
-Modified getassigned: /getassignedcourseids/ param: username/password to get course ids.
-Modified getassigned: /getassignedcourseids/ to return an xml with package and course details.
-Fixed bug in package upload . replaced all spaces in elp file with underscores.
-Seperated exe export and ui/cloud stuff. seperated Grunt tests as well.
-Fixed bug and now umcloud1 can recieve elp files from eXe 

v0.5.3
-Added coverage of unit tests in jenkins build script.
-Updates server steps to install coverage
-Added coverage to umclouddj-server-tar-deployment.sh
-Changed logged in message to show user's organisation and role at the top.
-Added icons to Management. 
-Moved Courses to Managemwent

lrs/views.py
import json
import logging
from django.conf import settings
from django.contrib.auth import authenticate, logout
from django.contrib.auth.decorators import login_required
from django.contrib.auth.models import User
from django.core.context_processors import csrf
from django.core.exceptions import ValidationError
from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
from django.core.urlresolvers import reverse
from django.http import HttpResponse, HttpResponseRedirect
from django.shortcuts import render_to_response
from django.template import RequestContext
from django.utils.decorators import decorator_from_middleware
from django.views.decorators.csrf import csrf_protect
from django.views.decorators.http import require_http_methods
from lrs import forms, models, exceptions
from lrs.util import req_validate, req_parse, req_process, XAPIVersionHeaderMiddleware, accept_middleware, StatementValidator
from oauth_provider.consts import ACCEPTED, CONSUMER_STATES

# This uses the lrs logger for LRS specific information
logger = logging.getLogger(__name__)
 
@login_required(login_url='/login/')
@decorator_from_middleware(accept_middleware.AcceptMiddleware)
@csrf_protect
def home(request):
    context = RequestContext(request)
    context.update(csrf(request))

    stats = {}
    stats['usercnt'] = User.objects.all().count()
    stats['stmtcnt'] = models.Statement.objects.all().count()
    stats['verbcnt'] = models.Verb.objects.all().count()
    stats['agentcnt'] = models.Agent.objects.filter().count()
    stats['activitycnt'] = models.Activity.objects.filter().count()
    return render_to_response('home.html', {'stats':stats}, context_instance=context)

@decorator_from_middleware(accept_middleware.AcceptMiddleware)
@csrf_protect
def stmt_validator(request):
    context = RequestContext(request)
    context.update(csrf(request))

    if request.method == 'GET':
        form = forms.ValidatorForm()
        return render_to_response('validator.html', {"form": form}, context_instance=context)
    elif request.method == 'POST':
        form = forms.ValidatorForm(request.POST)
        if form.is_valid():
            # Initialize validator (validates incoming data structure)
            try:
                validator = StatementValidator.StatementValidator(form.cleaned_data['jsondata'])
            except SyntaxError, se:
                return render_to_response('validator.html', {"form": form, "error_message": "Statement is not a properly formatted dictionary"},
                context_instance=context)
            except ValueError, ve:
                return render_to_response('validator.html', {"form": form, "error_message": "Statement is not a properly formatted dictionary"},
                context_instance=context)                
            except Exception, e:
                return render_to_response('validator.html', {"form": form, "error_message": e.message},
                context_instance=context)

            # Once know it's valid JSON, validate keys and fields
            try:
                valid = validator.validate()
            except exceptions.ParamError, e:
                return render_to_response('validator.html', {"form": form,"error_message": e.message},
                    context_instance=context)
            else:
                return render_to_response('validator.html', {"form": form,"valid_message": valid},
                    context_instance=context)
        else:
            return render_to_response('validator.html', {"form": form},
                context_instance=context)

@decorator_from_middleware(accept_middleware.AcceptMiddleware)
def about(request):
    lrs_data = { 
        "version": ["1.0.1"],
        "Extensions":{
            "xapi": {
                "statements":
                {
                    "name": "Statements",
                    "methods": ["GET", "POST", "PUT", "HEAD"],
                    "endpoint": reverse('lrs.views.statements'),
                    "description": "Endpoint to submit and retrieve XAPI statments.",
                },
                "activities":
                {
                    "name": "Activities",
                    "methods": ["GET", "HEAD"],
                    "endpoint": reverse('lrs.views.activities'),
                    "description": "Endpoint to retrieve a complete activity object.",
                },
                "activities_state":
                {
                    "name": "Activities State",
                    "methods": ["PUT","POST","GET","DELETE", "HEAD"],
                    "endpoint": reverse('lrs.views.activity_state'),
                    "description": "Stores, fetches, or deletes the document specified by the given stateId that exists in the context of the specified activity, agent, and registration (if specified).",
                },
                "activities_profile":
                {
                    "name": "Activities Profile",
                    "methods": ["PUT","POST","GET","DELETE", "HEAD"],
                    "endpoint": reverse('lrs.views.activity_profile'),
                    "description": "Saves/retrieves/deletes the specified profile document in the context of the specified activity.",
                },
                "agents":
                {
                    "name": "Agents",
                    "methods": ["GET", "HEAD"],
                    "endpoint": reverse('lrs.views.agents'),
                    "description": "Returns a special, Person object for a specified agent.",
                },
                "agents_profile":
                {
                    "name": "Agent Profile",
                    "methods": ["PUT","POST","GET","DELETE", "HEAD"],
                    "endpoint": reverse('lrs.views.agent_profile'),
                    "description": "Saves/retrieves/deletes the specified profile document in the context of the specified agent.",
                }
            },
            "lrs":{
                "user_register":
                {
                    "name": "User Registration",
                    "methods": ["POST"],
                    "endpoint": reverse('lrs.views.register'),
                    "description": "Registers a user within the LRS.",
                },
                "client_register":
                {
                    "name": "Client Registration",
                    "methods": ["POST"],
                    "endpoint": reverse('lrs.views.reg_client'),
                    "description": "Registers a client applicaton with the LRS.",
                }
            },
            "oauth":
            {
                "initiate":
                {
                    "name": "Oauth Initiate",
                    "methods": ["POST"],
                    "endpoint": reverse('oauth_provider.views.request_token'),
                    "description": "Authorize a client and return temporary credentials.",
                },
                "authorize":
                {
                    "name": "Oauth Authorize",
                    "methods": ["GET"],
                    "endpoint": reverse('oauth_provider.views.user_authorization'),
                    "description": "Authorize a user.",
                },
                "token":
                {
                    "name": "Oauth Token",
                    "methods": ["POST"],
                    "endpoint": reverse('oauth_provider.views.access_token'),
                    "description": "Provides Oauth token to the client.",
                }
            }
        }
    }    
    return HttpResponse(json.dumps(lrs_data), mimetype="application/json", status=200)

def actexample(request):
    return render_to_response('actexample.json', mimetype="application/json")

def actexample2(request):
    return render_to_response('actexample2.json', mimetype="application/json")

def actexample3(request):
    return render_to_response('actexample3.json', mimetype="application/json")

def actexample4(request):
    return render_to_response('actexample4.json', mimetype="application/json")

@csrf_protect
@require_http_methods(["POST", "GET"])
def register(request):
    context = RequestContext(request)
    context.update(csrf(request))
    
    if request.method == 'GET':
        form = forms.RegisterForm()
        return render_to_response('register.html', {"form": form}, context_instance=context)
    elif request.method == 'POST':
        form = forms.RegisterForm(request.POST)
        if form.is_valid():
            name = form.cleaned_data['username']
            pword = form.cleaned_data['password']
            email = form.cleaned_data['email']
            
            try:
                user = User.objects.get(username__exact=name)
            except User.DoesNotExist:
                user = User.objects.create_user(name, email, pword)
            else:
                return render_to_response('register.html', {"form": form, "error_message": "User %s already exists." % name},
                    context_instance=context)                
            
            d = {"info_message": "Thanks for registering. You can now use your name [%s] and password to sign in." % user.username}
            return render_to_response('reg_success.html', d, context_instance=context)
        else:
            return render_to_response('register.html', {"form": form}, context_instance=context)

@login_required(login_url="/XAPI/accounts/login")
@require_http_methods(["POST", "GET"])
def reg_client(request):
    if request.method == 'GET':
        form = forms.RegClientForm()
        return render_to_response('regclient.html', {"form": form}, context_instance=RequestContext(request))
    elif request.method == 'POST':
        form = forms.RegClientForm(request.POST)
        if form.is_valid():
            name = form.cleaned_data['name']
            description = form.cleaned_data['description']
            scopes = form.cleaned_data['scopes']
            
            try:
                client = models.Consumer.objects.get(name__exact=name)
            except models.Consumer.DoesNotExist:
                client = models.Consumer.objects.create(name=name, description=description, user=request.user,
                    status=ACCEPTED, default_scopes=",".join(scopes))
            else:
                return render_to_response('regclient.html', {"form": form, "error_message": "%s alreay exists." % name}, context_instance=RequestContext(request))         
            
            d = {"name":client.name,"app_id":client.key, "secret":client.secret, "info_message": "Your Client Credentials"}
            return render_to_response('reg_success.html', d, context_instance=RequestContext(request))
        else:
            return render_to_response('regclient.html', {"form": form}, context_instance=RequestContext(request))

@login_required(login_url="/XAPI/accounts/login")
def me(request):
    client_apps = models.Consumer.objects.filter(user=request.user)
    access_tokens = models.Token.objects.filter(user=request.user, token_type=models.Token.ACCESS, is_approved=True)
    return render_to_response('me.html', {'client_apps':client_apps, 'access_tokens':access_tokens},
        context_instance=RequestContext(request))

@login_required(login_url="/XAPI/accounts/login")
def my_statements(request):
    if request.method == "DELETE":
        models.Statement.objects.filter(user=request.user).delete()
        stmts = models.Statement.objects.filter(user=request.user)
        if not stmts:
            return HttpResponse(status=204)
        else:
            raise Exception("unable to delete statements")

    stmt_id = request.GET.get("stmt_id", None)
    if stmt_id:
        s = models.Statement.objects.get(statement_id=stmt_id, user=request.user)
        return HttpResponse(s.object_return(),mimetype="application/json",status=200)
    else:
        s = {}
        paginator = Paginator(models.Statement.objects.filter(user=request.user).order_by('-timestamp').values_list('id', flat=True), 
            settings.STMTS_PER_PAGE)

        page_no = request.GET.get('page', 1)
        try:
            page = paginator.page(page_no)
        except PageNotAnInteger:
            # If page is not an integer, deliver first page.
            page = paginator.page(1)
        except EmptyPage:
            # If page is out of range (e.g. 9999), deliver last page of results.
            page = paginator.page(paginator.num_pages)

        idlist = page.object_list
        if idlist.count() > 0:
            stmt_objs = [stmt for stmt in models.Statement.objects.filter(id__in=(idlist)).order_by('-timestamp')]
        else: 
            stmt_objs = []

        slist = []
        for stmt in stmt_objs:
            d = {}
            d['timestamp'] = stmt.timestamp.isoformat()
            d['statement_id'] = stmt.statement_id
            d['actor_name'] = stmt.actor.get_a_name()
            d['verb'] = stmt.verb.get_display()
            d['object'] = stmt.get_object().get_a_name()
            slist.append(d)

        s['stmts'] = slist
        if page.has_previous():
            s['previous'] = "%s?page=%s" % (reverse('lrs.views.my_statements'), page.previous_page_number())
        if page.has_next():
            s['next'] = "%s?page=%s" % (reverse('lrs.views.my_statements'), page.next_page_number())

        return HttpResponse(json.dumps(s), mimetype="application/json", status=200)

@login_required(login_url="/XAPI/accounts/login")
def my_app_status(request):
    try:
        name = request.GET['app_name']
        status = request.GET['status']
        new_status = [s[0] for s in CONSUMER_STATES if s[1] == status][0] #should only be 1
        client = models.Consumer.objects.get(name__exact=name, user=request.user)
        client.status = new_status
        client.save()
        ret = {"app_name":client.name, "status":client.get_status_display()}
        return HttpResponse(json.dumps(ret), mimetype="application/json", status=200)
    except:
        return HttpResponse(json.dumps({"error_message":"unable to fulfill request"}), mimetype="application/json", status=400)

@login_required(login_url="/XAPI/accounts/login")
@require_http_methods(["DELETE"])
def delete_token(request):
    try:
        ids = request.GET['id'].split("-")
        token_key = ids[0]
        consumer_id = ids[1]
        ts = ids[2]
        token = models.Token.objects.get(user=request.user,
                                         key__startswith=token_key,
                                         consumer__id=consumer_id,
                                         timestamp=ts,
                                         token_type=models.Token.ACCESS,
                                         is_approved=True)
        token.is_approved = False
        token.save()
        return HttpResponse("", status=204)
    except:
        return HttpResponse("Unknown token", status=400)

def logout_view(request):
    logout(request)
    # Redirect to a success page.
    return HttpResponseRedirect(reverse('lrs.views.home'))

# Called when user queries GET statement endpoint and returned list is larger than server limit (10)
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
@require_http_methods(["GET", "HEAD"])
def statements_more(request, more_id):
    return handle_request(request, more_id)

@require_http_methods(["PUT","GET","POST", "HEAD"])
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def statements(request):
    return handle_request(request)   

@require_http_methods(["PUT","POST","GET","DELETE", "HEAD"])
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def activity_state(request):
    return handle_request(request)  

@require_http_methods(["PUT","POST","GET","DELETE", "HEAD"])
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def activity_profile(request):
    return handle_request(request)

@require_http_methods(["GET", "HEAD"])
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def activities(request):
    return handle_request(request)

@require_http_methods(["PUT","POST","GET","DELETE", "HEAD"])    
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def agent_profile(request):
    return handle_request(request)

# returns a 405 (Method Not Allowed) if not a GET
@require_http_methods(["GET", "HEAD"])
@decorator_from_middleware(XAPIVersionHeaderMiddleware.XAPIVersionHeader)
def agents(request):
    return handle_request(request)

@login_required
def user_profile(request):
    return render_to_response('registration/profile.html')

validators = {
    reverse(statements).lower() : {
        "POST" : req_validate.statements_post,
        "GET" : req_validate.statements_get,
        "PUT" : req_validate.statements_put,
        "HEAD" : req_validate.statements_get
    },
    reverse(activity_state).lower() : {
        "POST": req_validate.activity_state_post,
        "PUT" : req_validate.activity_state_put,
        "GET" : req_validate.activity_state_get,
        "HEAD" : req_validate.activity_state_get,
        "DELETE" : req_validate.activity_state_delete
    },
    reverse(activity_profile).lower() : {
        "POST": req_validate.activity_profile_post,
        "PUT" : req_validate.activity_profile_put,
        "GET" : req_validate.activity_profile_get,
        "HEAD" : req_validate.activity_profile_get,
        "DELETE" : req_validate.activity_profile_delete
    },
    reverse(activities).lower() : {
        "GET" : req_validate.activities_get,
        "HEAD" : req_validate.activities_get
    },
    reverse(agent_profile).lower() : {
        "POST": req_validate.agent_profile_post,
        "PUT" : req_validate.agent_profile_put,
        "GET" : req_validate.agent_profile_get,
        "HEAD" : req_validate.agent_profile_get,
        "DELETE" : req_validate.agent_profile_delete
    },
   reverse(agents).lower() : {
       "GET" : req_validate.agents_get,
       "HEAD" : req_validate.agents_get
   },
   "/xapi/statements/more" : {
        "GET" : req_validate.statements_more_get,
        "HEAD" : req_validate.statements_more_get
   }
}

processors = {
    reverse(statements).lower() : {
        "POST" : req_process.statements_post,
        "GET" : req_process.statements_get,
        "HEAD" : req_process.statements_get,
        "PUT" : req_process.statements_put
    },
    reverse(activity_state).lower() : {
        "POST": req_process.activity_state_post,
        "PUT" : req_process.activity_state_put,
        "GET" : req_process.activity_state_get,
        "HEAD" : req_process.activity_state_get,
        "DELETE" : req_process.activity_state_delete
    },
    reverse(activity_profile).lower() : {
        "POST": req_process.activity_profile_post,
        "PUT" : req_process.activity_profile_put,
        "GET" : req_process.activity_profile_get,
        "HEAD" : req_process.activity_profile_get,
        "DELETE" : req_process.activity_profile_delete
    },
    reverse(activities).lower() : {
        "GET" : req_process.activities_get,
        "HEAD" : req_process.activities_get
    },
    reverse(agent_profile).lower() : {
        "POST": req_process.agent_profile_post,
        "PUT" : req_process.agent_profile_put,
        "GET" : req_process.agent_profile_get,
        "HEAD" : req_process.agent_profile_get,
        "DELETE" : req_process.agent_profile_delete
    },
   reverse(agents).lower() : {
       "GET" : req_process.agents_get,
       "HEAD" : req_process.agents_get
   },
   "/xapi/statements/more" : {
        "GET" : req_process.statements_more_get,
        "HEAD" : req_process.statements_more_get
   }      
}

def handle_request(request, more_id=None):
    try:
        r_dict = req_parse.parse(request, more_id)
        path = request.path.lower()

        if path.endswith('/'):
            path = path.rstrip('/')

        # Cutoff more_id
        if '/xapi/statements/more' in path:
            path = '/xapi/statements/more'

        req_dict = validators[path][r_dict['method']](r_dict)
        return processors[path][req_dict['method']](req_dict)

    except exceptions.BadRequest as err:
        log_exception(request.path, err)
        return HttpResponse(err.message, status=400)
    except ValidationError as ve:
        log_exception(request.path, ve)
        return HttpResponse(ve.messages[0], status=400)
    except exceptions.Unauthorized as autherr:
        log_exception(request.path, autherr)
        r = HttpResponse(autherr, status = 401)
        r['WWW-Authenticate'] = 'Basic realm="ADLLRS"'
        return r
    except exceptions.OauthUnauthorized as oauth_err:
        log_exception(request.path, oauth_err)
        return oauth_err.response
    except exceptions.Forbidden as forb:
        log_exception(request.path, forb)
        return HttpResponse(forb.message, status=403)
    except exceptions.NotFound as nf:
        log_exception(request.path, nf)
        return HttpResponse(nf.message, status=404)
    except exceptions.Conflict as c:
        log_exception(request.path, c)
        return HttpResponse(c.message, status=409)
    except exceptions.PreconditionFail as pf:
        log_exception(request.path, pf)
        return HttpResponse(pf.message, status=412)
    except Exception as err:
        log_exception(request.path, err)
        return HttpResponse(err.message, status=500)

def log_exception(path, ex):
    logger.info("\nException while processing: %s" % path)
    logger.exception(ex)
